dbms/src/DataStreams/MergeSortingBlockInputStream.cpp

```diff
@@ -125,41 +126,33 @@ Block MergeSortingBlockInputStream::readImpl()
               */
             if (max_bytes_before_external_sort && sum_bytes_in_blocks > max_bytes_before_external_sort)
             {
-                temporary_files.emplace_back(new Poco::TemporaryFile(tmp_path));
-                const std::string & path = temporary_files.back()->path();
-                WriteBufferFromFile file_buf(path);
-                CompressedWriteBuffer compressed_buf(file_buf);
-                NativeBlockOutputStream block_out(compressed_buf, 0, header_without_constants);
                 MergeSortingBlocksBlockInputStream block_in(blocks, description, log->identifier(), max_merged_block_size, limit);
-
-                LOG_INFO(log, "Sorting and writing part of data into temporary file {}", path);
-                copyData(block_in, block_out, &is_cancelled); /// NOTE. Possibly limit disk usage.
-                LOG_INFO(log, "Done writing part of data into temporary file {}", path);
-
+                auto all_blocks = readAllData(block_in, &is_cancelled);
```

Comment:
Is there a sub-batch to be considered here?
Feels like there is a risk of oom.

Reply:
> Is there a sub-batch to be considered here? Feels like there is a risk of oom.
MergeSortingBlocksBlockInputStream holds reference of blocks, readAllData will handle all blocks and move to all_blocks, I think it may hold at most the memory used by blocks.

Reply:
> Is there a sub-batch to be considered here? Feels like there is a risk of oom.

MergeSortingBlocksBlockInputStream holds reference of blocks, readAllData will handle all blocks and move to all_blocks, I think it may hold at most the memory used by blocks.

Reply:
The peak memory usage after the pr will be `2 * bytes_of(blocks)`, and before this pr the peak memory usage is `bytes_of(blocks) + bytes_of(block)`. In order to avoid oom, I will split the spill into multiple rounds, each round, the spilled data is `max_bytes`, so the peak memory usage will reduced to `bytes_of(blocks) + max_bytes`.

Reply:
I've added a new spill function to split the spill into multiple rounds.